File: study-algorithm/src/nedaair/study/sort/quickSort.py
__author__ = 'nedaair'



unSort = [31, 8, 48, 73, 11, 3, 20, 29, 65, 15]

def quickSort(unSort, p, r) :

    if p < r :
        q = partition(unSort, p, r)
        quickSort(unSort, p, q-1)
        quickSort(unSort, q+1, r)
    pass


def partition(unSort, p, r) :

    x = unSort[r]
    i = p -1
    for j in range(p, r) :
        if unSort[j] < x :
            i += 1

            temp = unSort[i]
            unSort[i] = unSort[j]
            unSort[j] = temp


    unSort[r] = unSort[i+1]
    unSort[i+1] = x

    return i+1



quickSort(unSort, 0, len(unSort)-1)


print unSort

File: study-algorithm/src/xenomity/examination/sort/QuickSort.java
package xenomity.examination.sort;


/**
 * Quick Sort
 * 
 * @author Xenomity a.k.a P-slinc' (이승한)
 * @since 2015. 1. 31.
 *
 */
public class QuickSort implements Sortable {

	@Override
	public void sort(int[] array, int... args) {
		// definition
		int startIndex = (args != null && args.length == 2) ? args[0] : 0;
		int endIndex = (args != null && args.length == 2) ? args[1] : array.length - 1;
		
		if (startIndex < endIndex) {
			// #1. pivot selection
			int pivot = partition(array, startIndex, endIndex);
			
			// #2. recursive sort
			sort(array, startIndex, pivot - 1);
			sort(array, pivot + 1, endIndex);
		}
	}
	
	private int partition(int[] array, int startIndex, int endIndex) {
		int criteriaValue = array[endIndex];
		int pivotIndex = startIndex;
		
		// 순회하면서 피봇 값보다 작은 위치가 나올 경우마다 피봇 인덱스를 증가시키며 현재 원소와 피봇 위치 원소 교환.
		for (int i = startIndex; i < endIndex; i++) {
			if (array[i] <= criteriaValue) {
				int temp = array[pivotIndex];
				array[pivotIndex] = array[i];
				array[i] = temp;
				
				pivotIndex++;
			}
		}
		
		// 최종적인 피봇 위치의 원소와 기준 원소를 교환.
		array[endIndex] = array[pivotIndex];
		array[pivotIndex] = criteriaValue;
		
		return pivotIndex;
	}
	
}
